vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/AbstractNumberField.java

```diff
@@ -407,4 +427,164 @@ protected void onAttach(AttachEvent attachEvent) {
         super.onAttach(attachEvent);
         ClientValidationUtil.preventWebComponentFromModifyingInvalidState(this);
     }
+
+    /**
+     * Gets the internationalization object previously set for this component.
+     * <p>
+     * Note: updating the object content that is gotten from this method will
+     * not update the lang on the component if not set back using
+     * {@link AbstractNumberField#setI18n(AbstractNumberFieldI18n)}
+     *
+     * @return the i18n object. It will be {@code null}, If the i18n properties
+     *         weren't set.
+     */
+    protected AbstractNumberFieldI18n getI18n() {
+        return i18n;
+    }
+
+    /**
+     * Sets the internationalization properties for this component.
+     *
+     * @param i18n
+     *            the internationalized properties, not {@code null}
+     */
+    protected void setI18n(AbstractNumberFieldI18n i18n) {
+        this.i18n = Objects.requireNonNull(i18n, "The i18n properties object should not be null");
+    }
+
+    private String getBadInputErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(AbstractNumberFieldI18n::getBadInputErrorMessage).orElse("");
+    }
+
+    private String getRequiredErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(AbstractNumberFieldI18n::getRequiredErrorMessage).orElse("");
+    }
+
+    private String getMinErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(AbstractNumberFieldI18n::getMinErrorMessage).orElse("");
+    }
+
+    private String getMaxErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(AbstractNumberFieldI18n::getMaxErrorMessage).orElse("");
+    }
+
+    private String getStepErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(AbstractNumberFieldI18n::getStepErrorMessage).orElse("");
+    }
+
+    /**
+     * The abstract internationalization properties for {@link AbstractNumberField}.
+     */
+    public static abstract class AbstractNumberFieldI18n
+            implements Serializable {
```

Comment:
What are the benefits of the abstract class if the implementing class has to do all the "heavily" lifting? (Like overriding all methods)

Reply:
Mmm, right. I think I could avoid needing an abstract class by making the visibility of getRequiredErrorMessage-like methods package-private and overriding them in the implementing classes to return error messages from their i18n objects. Not sure if it's the best alternative, but this should at least save us from having this abstract class which is public.

Reply:
Just asking without much thought: Wouldn't it be possible to create a shared instance `NumberFieldI18n` which is used in all different number fields? It feels kinda overwhelming to e.g. maintain different classes for a field "type" that's so similar.

Reply:
I brought this up in our internal discussion and we leaned toward proceeding with separate i18n classes for each number field class to keep it consistent with other fields. Using a single class could lead to breaking changes in the setI18n / getI18n signatures if we later need to add properties specific to a single number field, even though this is unlikely. Given that we already use AbstractNumberField, using AbstractNumberFieldI18n feels more consistent. Package-private methods are another option, but the abstract class seems more appropriate.

Reply:
As a debate: I wondered why this class couldn't have a generic type like `<T extends AbstractNumberFieldI18n<T>>` and then use it as the return type of the setters. From what I can see, the current contract could allow a child class to return an instance of a second child class from a setter.

Reply:
We talked about this approach internally and found that the developer can still omit the generic <T> parameter. Without this parameter, there is no limit on the return type. So, the approach doesn't offer much benefit after all.